### PR TITLE
Create tabular/rawsotre rest api parameters

### PR DESCRIPTION
**Pull request by zzbdatatom**
PR checklist:
- Did you run ClangFormat ? yes
- Did you run CheckStyle ? yes
- Did you check the comment / log / exception conventions in Engineering code process wiki page ? yes
- Did you import unnecessary headers ? no
- Did you surround `Daisy : starts/ends` for new code in existing ClickHouse code base ? yes

Please write user-readable short description of the changes:
1. Support the parsing and setting of the uri parameter  in crate tabular/rawsotre
2. Replace executeQuery with executeSelectQuery in DatabaseRestRouterHandler
3. Fix the problem of duplicate table in table list cluster mode

### CODE REVIEW

File: src/Server/RestRouterHandlers/TableRestRouterHandler.cpp [hunk 1]
```diff
@@ -277,6 +290,24 @@ String TableRestRouterHandler::getCreationSQL(const Poco::JSON::Object::Ptr & pa
...
+    for (const auto & setting : CREATE_TABLE_SETTINGS)
+    {
+        if (hasQueryParameter(setting))
+        {
+            const String & setting_str = setting + "=" + getQueryParameter(setting);
+            create_segments.push_back(", " + setting_str);
+
+            if (!create_table_settings.empty())
+            {
+                create_table_settings += "&";
```
Comment (hzhzhang) [inline src/Server/RestRouterHandlers/TableRestRouterHandler.cpp]: Suggest to use boost::algorithm::join() here

Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/TableRestRouterHandler.cpp]: Good suggestion

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp [hunk 2]
```diff
@@ -30,7 +30,8 @@ std::pair<String, Int32> DatabaseRestRouterHandler::executeGet(const Poco::JSON:
 {
     String query = "SHOW DATABASES;";
 
-    return processQuery(query);
+    Poco::JSON::Object resp;
+    return {processQuery(query, ([this, &resp](Block && block) { this->buildDatabaseArray(block, resp); }), resp), HTTPResponse::HTTP_OK};
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp]: This line is a bit complex, can we separate `processQuery` call to a separate line ?

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.h [hunk 3]
```diff
@@ -20,9 +20,7 @@ class DatabaseRestRouterHandler final : public RestRouterHandler
     std::pair<String, Int32> executeDelete(const Poco::JSON::Object::Ptr & payload) const override;
 
 private:
-    std::pair<String, Int32> processQuery(const String & query) const;
-    String buildResponse() const;
-    void processQueryWithProcessors(Poco::JSON::Object & resp, QueryPipeline & pipeline) const;
+    void buildDatabaseArray(const Block & block, Poco::JSON::Object & resp) const;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/DatabaseRestRouterHandler.h]: This can be purely static free function instead of a member function ? Personally i prefer static free function in anonymous namespace if the function is not reused by others

File: src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp [hunk 4]
```diff
@@ -31,8 +31,13 @@ void RawstoreTableRestRouterHandler::buildTablesJSON(Poco::JSON::Object & resp,
 {
     Poco::JSON::Array tables_mapping_json;
 
+    std::vector<String> table_names;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp]: we can use a `std::unordered_set<String>` ?

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp [hunk 5]
```diff
@@ -86,4 +98,12 @@ void RestRouterHandler::setupDistributedQueryParameters(
     }
     query_context->setDistributedDDLOperation(true);
 }
+
+String
+RestRouterHandler::processQuery(const String & query, const std::function<void(Block &&)> & callback, const Poco::JSON::Object & resp) const
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.cpp]: Let's accept `Poco::JSON::Object & resp` and `std::function<void(Block &&)> callback`

File: src/Server/RestRouterHandlers/TableRestRouterHandler.cpp [hunk 6]
```diff
@@ -277,6 +290,19 @@ String TableRestRouterHandler::getCreationSQL(const Poco::JSON::Object::Ptr & pa
...
+    {
+        if (hasQueryParameter(setting))
+        {
+            const String & setting_str = setting + "=" + getQueryParameter(setting);
+
+            create_segments.push_back(", " + setting_str);
+            create_table_settings.push_back(setting_str);
+        }
+    }
+    query_context->setQueryParameter("create_table_settings", boost::algorithm::join(create_table_settings, "&"));
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/TableRestRouterHandler.cpp]: Only when `distributed_ddl=true`, we will need setup query parameter in context ? Actually, can we bindly setup all URL parameters in one shot : is it possible to get URL like `a=b&c=d&e=f...` and then setup `query_context->setQueryParameter("url_paramaters", url_params)` ? Then we don't need introduce different names for different url params ?

File: src/Server/RestRouterHandlers/TabularTableRestRouterHandler.cpp [hunk 7]
```diff
@@ -49,9 +49,14 @@ std::map<String, std::map<String, String> > TabularTableRestRouterHandler::colum
 void TabularTableRestRouterHandler::buildTablesJSON(Poco::JSON::Object & resp, const CatalogService::TablePtrs & tables) const
 {
     Poco::JSON::Array tables_mapping_json;
+    std::vector<String> table_names;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/TabularTableRestRouterHandler.cpp]: `unordered_set<String>` can serve your dedup purpose better

File: src/Server/RestRouterHandlers/TableRestRouterHandler.cpp (hunk 6 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/TableRestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp (hunk 2 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.h (hunk 3 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/DatabaseRestRouterHandler.h]: fixed

File: src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp (hunk 4 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp (hunk 5 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/TabularTableRestRouterHandler.cpp (hunk 7 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/TabularTableRestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp [hunk 8]
```diff
@@ -13,12 +10,27 @@ namespace DB
...
+        {"required",{
+                        {"name", "string"}
+                    }
+        }
+    };
+
+    void buildDatabaseArray(const Block & block, Poco::JSON::Object & resp)
+    {
+        Poco::JSON::Array databases_mapping_json;
+        if (block)
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp]: we don't need this checking. if it is empty, `block.rows()` will be 0 ?

File: src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp [hunk 9]
```diff
@@ -30,12 +30,11 @@ std::map<String, std::map<String, String> > RawstoreTableRestRouterHandler::crea
...
 {
     Poco::JSON::Array tables_mapping_json;
-
-    std::vector<String> table_names;
+    std::unordered_set<String> table_names;
 
     for (const auto & table : tables)
     {
-        if (std::find(table_names.begin(), table_names.end(), table->name) != table_names.end())
+        if (table_names.find(table->name) != table_names.end())
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp]: `table_names.contains(...)`

Comment (chenziliang) [reply src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp]: Can we calculate `shards` and `replication_factor` for table list ? We can do this in a different PR

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp [hunk 10]
```diff
@@ -96,11 +96,20 @@ void RestRouterHandler::setupDistributedQueryParameters(
     {
         query_context->setQueryParameter(kv.first, kv.second);
     }
+
+    /// setup query parameter in context : a=b&c=d&e=f
+    std::vector<String> url_params;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.cpp]: nit: can we have a different way to get the raw URL parameters like from the HTTPRequest ? we can save a copy of the raw URL parameters in base class if necessary

File: src/Server/RestRouterHandlers/RestRouterHandler.h [hunk 11]
```diff
@@ -119,8 +119,8 @@ class RestRouterHandler : private boost::noncopyable
 
     String processQuery(
         const String & query,
-        const std::function<void(Block &&)> & callback = [](Block &&) {},
-        const Poco::JSON::Object & resp = Poco::JSON::Object()) const;
+        const Poco::JSON::Object & resp = Poco::JSON::Object(),
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.h]: Let's accept `Poco::JSON::Object & resp`.  We can have overloaded `processQuery` functions if necessary. For example, if sometimes we don't care `resp`, we can have `processQuery(const String & query, std::function<void(Block &&)> callback)`; if user don't care `callback`, then we can have a different one ...

File: src/Server/RestRouterHandlers/TableRestRouterHandler.cpp [hunk 12]
```diff
@@ -290,18 +290,13 @@ String TableRestRouterHandler::getCreationSQL(const Poco::JSON::Object::Ptr & pa
...
-    std::vector<String> create_table_settings;
     for (const auto & setting : CREATE_TABLE_SETTINGS)
     {
         if (hasQueryParameter(setting))
         {
-            const String & setting_str = setting + "=" + getQueryParameter(setting);
-
-            create_segments.push_back(", " + setting_str);
-            create_table_settings.push_back(setting_str);
+            create_segments.push_back(", " + setting+ "=" + getQueryParameter(setting));
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/TableRestRouterHandler.cpp:298]: Ideally we shall do some parameters validation like the settings are not negative integers etc. We can do this in a different PR along with `shards` and `replication_factor` calculation if you like

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp (hunk 10 above)
Comment (chenziliang) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: Because simply k=v concatenation sometimes doesn't work.

File: src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp (hunk 9 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RawstoreTableRestRouterHandler.cpp]: Yes, I will build a related task

File: src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp (hunk 8 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/DatabaseRestRouterHandler.cpp]: fixed

File: src/Server/RestRouterHandlers/TableRestRouterHandler.cpp (hunk 12 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/TableRestRouterHandler.cpp:298]: Related fixme has been added

File: src/Server/RestRouterHandlers/RestRouterHandler.h (hunk 11 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.h]: fixed

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp (hunk 10 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: Sorry, I don’t really understand here， URL parameters are parsed from HTTPRequest  even it doesn't work.

Comment (chenziliang) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: url parameters shall be escaped sometimes, like there is `=` sign at the beginning of value. Simple concatenation something like k + '=' + v would result in something like `k==abc` if value itself is `=abc`.

File: src/Server/RestRouterHandlers/RestRouterHandler.h [hunk 13]
```diff
@@ -118,9 +118,10 @@ class RestRouterHandler : private boost::noncopyable
     }
 
     String processQuery(
-        const String & query,
-        const Poco::JSON::Object & resp = Poco::JSON::Object(),
-        const std::function<void(Block &&)> & callback = [](Block &&) {}) const;
+        const String & query, const std::function<void(Block &&)> & callback = [](Block &&) {}) const;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.h:121]: Can we use `std::function<void(Block &&)> callback` as the parameter ?

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp (hunk 10 above)
Comment (chenziliang) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: HTTPRequest::getURI()::getRawQuery() may be something you like to explore

File: src/Server/RestRouterHandlers/RestRouterHandler.h (hunk 13 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.h:121]: No we can't, because in executeSelectQuery
In, callbacak is also modified with constant, That is : executeSelectQuery(const String & query, ContextPtr query_context, const std::function<void(Block &&)> & callback, bool internal)

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp [hunk 14]
```diff
@@ -58,6 +58,11 @@ void RestRouterHandler::execute(HTTPServerRequest & request, HTTPServerResponse
         }
         else
         {
+            if (isDistributedDDL())
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.cpp]: can we do this setup in `setupHTTPContext` ?

File: src/Server/RestRouterHandlers/RestRouterHandler.h [hunk 15]
```diff
@@ -213,6 +213,13 @@ class RestRouterHandler : private boost::noncopyable
 
     void setupQueryParams(const HTTPServerRequest & request) { query_parameters = std::make_unique<HTMLForm>(request); }
 
+    void setupRawQuery(const HTTPServerRequest & request)
+    { 
+        Poco::URI uri(request.getURI());
+        const std::string & query = uri.getRawQuery();
+        query_context->setQueryParameter("url_paramaters", query);
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.h]: `query_context->setQueryParameter("url_paramaters", uri.getRawQuery());`

File: src/Server/RestRouterHandlers/RestRouterHandler.h [hunk 16]
```diff
@@ -117,6 +117,12 @@ class RestRouterHandler : private boost::noncopyable
         return jsonErrorResponse("Internal server error", error_code);
     }
 
+    String processQuery(
+        const String & query, const std::function<void(Block &&)> & callback = [](Block &&) {}) const;
+
+    String processQuery(
+        const String & query, Poco::JSON::Object & resp, const std::function<void(Block &&)> & callback = [](Block &&) {}) const;
```
Comment (chenziliang) [inline src/Server/RestRouterHandlers/RestRouterHandler.h]: do we need give a default value to `callback` ?

Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.h]: fixed

File: src/Server/RestRouterHandlers/RestRouterHandler.h (hunk 15 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.h]: fixed

File: src/Server/RestRouterHandlers/RestRouterHandler.cpp (hunk 14 above)
Comment (zzbdatatom) [reply src/Server/RestRouterHandlers/RestRouterHandler.cpp]: Yes

**Review by chenziliang — APPROVED** [review]
